common/flatpak-dir.c

```diff
@@ -3434,6 +3467,36 @@ upgrade_deploy_data (GBytes *deploy_data,
                              g_variant_new_variant (g_variant_new_uint64 (0)));
     }
 
+  /* Deploy versions older than 4 might have some of the below fields, but it's
+   * not guaranteed if the deploy was first created with an old Flatpak version
+   */
+  if (old_version < 4)
+    {
+      const char *commit;
+      g_autoptr(GVariant) commit_data = NULL;
+      g_autoptr(GVariant) commit_metadata = NULL;
+      g_autoptr(GKeyFile) keyfile = NULL;
+      g_autoptr(GFile) metadata_file = NULL;
+      g_autofree char *metadata_contents = NULL;
+
+      /* Add fields from commit metadata to deploy */
+      commit = flatpak_deploy_data_get_commit (deploy_data);
+      if (!ostree_repo_load_commit (repo, commit, &commit_data, NULL, error))
+        return NULL;
+      commit_metadata = g_variant_get_child_value (commit_data, 0);
+      add_commit_metadata_to_deploy_data (&metadata_builder, commit_metadata);
```

Comment:
Doesn't this duplicate all values that were already added to metadata_builder? I think we need to use a GVariantDict instead of a GVariantBuilder to avoid problems with this.

Reply:
Good point